Added optional ALSA support

Siborgium opened the pull request:
Both PulseAudio and ALSA modules are optional (enabled by default). Further, they properly work together (as far as I could test it on my machine).

Siborgium committed:
Added ALSA support
Both PulseAudio and ALSA modules are optional (enabled by default)

Siborgium committed:
Removed old sound.rs

kennylevinsen commented on Cargo.toml:
```diff
@@ -8,7 +8,11 @@ edition = "2018"
 [features]
+default = ["alsasound", "pulseaudio"]
 damage_debug = []
+alsasound = ["alsa"]
```
This isn't consistent with the naming of the "pulseaudio" feature. I'd just name it `alsa` for now.

Siborgium replied:
However, it is consistent with the naming of the libraries it depends on. If rename `alsasound`, I'd stick to `asound`. It is not possible to rename the feature to`alsa`, because the dependency is called `alsa` itself. We can rename it though, but I don't think it is a good idea.

kennylevinsen replied:
Maybe "alsa-widget" and "pulseaudio-widget"?

We shouldn't rename the deps, but the feature names should be descriptive and consistent with regards to what functionality they provide (not what additional dependencies they cause).

Siborgium replied:
Well, for me the only that matters is the dependency. I avoid using pulseaudio and I don't wish to have it in my packages. It is possible to put the list of features and their dependencies in README.md, like
```md
* alsa-widget:  alsa-lib
* pulseaudio-widget: libpulse
```

By the way, since pulseaudio depends on `libpulse`, it could be `pulse-audio` and `alsa-audio`.

kennylevinsen replied:
Feature flags control _features_, not _dependencies_. That they spare you of dependencies is an indirect (positive) side-effect.

"-audio" is technically fine, but the proper name is "pulseaudio", and "pulseaudio-audio" sounds silly. I think "alsa-widget" and "pulseaudio-widget" is fine.

Siborgium replied:
*For me*. I don't want my personal build to depend on pulseaudio, and it is far easier to check what it depends on if features are named after the dependencies they require.

But, after all, you're the author, so be it, I'll change it.

kennylevinsen replied:
Thank you.

Note that I am sympathetic to not wanting dependencies on PulseAudio (which is why I'm happy to merge this PR), I just disagree with the purpose of a feature flag.

kennylevinsen commented on src/widgets/alsa_sound.rs:
```diff
@@ -0,0 +1,138 @@
+    pub mixer: Mixer,
+}
+
+//unsafe impl Send for AlsaMixer {}
```
Dead - Please remove. (Also, this is usually a bad idea to do.)

kennylevinsen commented on src/widgets/alsa_sound.rs:
```diff
@@ -0,0 +1,138 @@
+}
+
+pub struct AlsaVolume {
+    inner: Arc<Mutex<AlsaMixer>>
```
Why is this `Arc<Mutex<T>>` when there are no threads in the implementation? (When removed, using an inner type also ends up redundant.)

kennylevinsen commented on src/widgets/alsa_sound.rs:
```diff
@@ -0,0 +1,138 @@
+}
+
+impl AlsaVolume {
+    pub fn initialize() -> ::std::io::Result<Self> {
```
`initialize` seems redundant. Just do this directly in `new`.

kennylevinsen commented on src/widgets/alsa_sound.rs:
```diff
@@ -0,0 +1,138 @@
+    }
+    pub fn get_master_volume(&self) -> ::std::io::Result<i64> {
+        self.inner
+            .try_lock()
```
I've written elsewhere that the lock is redundant, but using `try_lock` is definitely wrong. If the lock had been necessary, then it would make the getters/setters here fail arbitrarily.

If the lock had been necessary, then `let inner = self.inner.lock().unwrap();` would pretty much be the only correct way to use it (including the unwrap - the lock will only fail if another thread has poisoned it by panicking, to which the only sane response 99% of the time is to panic as well).

kennylevinsen commented on src/widgets/alsa_sound.rs:
```diff
@@ -0,0 +1,138 @@
+}
+
+
+// Note: ALSA returns values from 0 to 65536
```
This note doesn't make sense here.

kennylevinsen commented on src/widgets/alsa_sound.rs:
```diff
@@ -0,0 +1,138 @@
+        "volume"
+    }
+    fn value(&self) -> f32 {
+        // TODO: use min/max instead of hardcoded constants
```
Elaborate? Is this something that needs to be done or not?

Siborgium committed:
Use `playback_volume_range` instead of hardcoded constants
Less amount of `Master` lookups per call
Alsa-related functions receive and return floats (in range of 0.0f32-1.0f32)

kennylevinsen commented on src/widgets/alsa_sound.rs:
```diff
@@ -0,0 +1,138 @@
+            .unwrap_or_else(|e| { eprintln!("{}", e); 0.0f32 })
+    }
+    fn color(&self) -> Color {
+        // TODO: Custom color from config
```
This isn't ALSA-specific, and should be a Github issue rather than a TODO.

kennylevinsen commented on src/widgets/alsa_sound.rs:
```diff
@@ -0,0 +1,138 @@
+            })
+            .and_then(|ref mixer| {
+                let volume = mixer.get_master_volume()
+                    .map_err(|e| eprintln!("{}", e))
```
Don't do random prints. You should probably just ? get_master_volume.

Siborgium replied:
It is not possible, because function returns `f32`, not `std::io::Result<f32>`. I can just `unwrap_or`, but silent errors are not good, panicking is not good either because user wouldn't like launcher to fail when some of its parts don't work.

kennylevinsen replied:
This closure returns a `Result`.

Furthermore, depending on the situation, failing as a result of misconfiguration is acceptable. A call to `get_master_volume` in `new` could also be used to probe availability, removing the widget if it won't work at all.

Siborgium replied:
Yes, it does, but `inc` returns `f32`. The fail in this case is not a result of misconfiguration, it is an IO error. Furthermore, if call to `get_master_volume` succeeds once, it, strictly speaking, is not guaranteed to succeed in the future. However, if it succeeds once, there is almost 100% chance it will succeed again, so `Err` branch won't be triggered often. Honestly, I couldn't even manage to trigger it once.
Probing sounds nice, I'll add it, but I still see no reason to neither remove error info nor panic on failed `inc`.

kennylevinsen replied:
I'm OK with this PR as is after the feature is renamed, but:

> Yes, it does, but inc returns f32

I was referring to returning a result from the closure, which you already did for set_master_volume, to reuse your existing error handling rather than just printing an error and defaulting to 0, especially seeing that it's unlikely that set_master_volume will succeed if get_master_volume had failed.

For the error handling, the errors are unlikely to occur if it works once, which is also why I think having non-panic error handling is redundant as is. unwrap/expect is easier to read than nested and_then.

I am, however, not very fond of unstructured log output, which is just noise. There's no log facility in place, so widget errors are either that the widget does not appear (failed probe), or that wldash panics (which notifies the user that something is broken). Printing unstructured data to stderr that no one will see (starting wldash from an interactive terminal to monitor output is not a normal use-case), is less helpful than a panic.

kennylevinsen commented on src/widgets/alsa_sound.rs:
```diff
@@ -0,0 +1,138 @@
+                    .unwrap_or(0) as f32;
+                let volume = volume + diff * 65536.0f32;
+                let volume = if volume > 65536.0f32 { 65536 } else { volume as i64 };
+                eprintln!("Inc: {}", volume);
```
Don't do random prints.

kennylevinsen commented on src/widgets/alsa_sound.rs:
```diff
@@ -0,0 +1,138 @@
+
```
Could we make a widgets/audio module and put alsa and pulseaudio in there, so the modules are at `widgets::audio::Alsa` and `widgets::audio::PulseAudio`? That would be nice. :)

kennylevinsen reviewed (changes requested):
Thanks!

Don't get scared by the number of comments: It's all small stuff (e.g. "remove dead comment"/"remove random prints").

One thing was that it would be nice to have pulseaudio/alsa together in an `audio` module folder (I named it `sound` before, but `audio` is definitely better).

Siborgium committed:
All requested changes (except for those I'd like to discuss) were
implemented:
Both ALSA and PulseAudio were placed in `audio` module
Cleaned code, removed dead comments
Removed AlsaVolume layer, renamed AlsaMixer to Alsa
Added efficient `inc` implementation

kennylevinsen reviewed (changes requested):
Looks good - will merge when the feature names are in place.

Siborgium committed:
Minor cleanup

Siborgium committed:
Features are named as requested (pulseaudio-widget, alsa-widget)
Probing ALSA on startup
panic! in case of ALSA error instead of logging to stderr

Siborgium committed:
Missing preprocessor directive

kennylevinsen commented:
Thanks, and sorry for my picky review! I'll try to be gentler in the future. :)
